File: src/core/java/src/net/lintim/model/impl/FullOD.java
package net.lintim.model.impl;

import net.lintim.model.OD;
import net.lintim.model.ODPair;

import java.util.Arrays;
import java.util.List;
import java.util.Objects;
import java.util.stream.Collectors;

/**
 * Implements the od interface using a two dimensional array. Therefore accessing the od pairs by index is cheap, but
 * getting nonempty od-pairs by {@link #getODPairs()} may be expensive.
 */
public class FullOD implements OD {
    /**
     * The data object to store the od pairs in. Remember to use index shifting for every raw access inside this class!
     */
    private ODPair[][] matrix;

    /**
     * Create a new full od matrix. This matrix will store the od pairs in a two dimensional array. Therefore accessing
     * the od pairs by index is cheap, but getting nonempty od-pairs by {@link #getODPairs()} may be expensive.
     * @param size the size of the matrix
     */
    public  FullOD(int size){
        this.matrix = new ODPair[size][size];
    }
    @Override
    public double getValue(int origin, int destination) {
        throwForInvalidIndexPair(origin, destination);
        return matrix[origin-1][destination-1] != null ? matrix[origin-1][destination-1].getValue() : 0;
    }

    @Override
    public void setValue(int origin, int destination, double newValue) {
        throwForInvalidIndexPair(origin, destination);
        matrix[origin-1][destination-1] = new ODPair(origin, destination, newValue);
    }

    @Override
    public double computeNumberOfPassengers() {
        int sum = 0;
        for(int origin = 1; origin <= matrix.length; origin++){
            for(int destination = 1; destination <= matrix.length; destination++){
                sum += getValue(origin, destination);
            }
        }
        return sum;
    }

    @Override
    public List<ODPair> getODPairs() {
        return Arrays.stream(matrix) //Stream of arrays
            .flatMap(Arrays::stream) //Stream of od pairs
            .filter(Objects::nonNull) //May be null
            .filter(odPair -> odPair.getValue() > 0) //Choose only nonempty od pairs
            .collect(Collectors.toList());
    }

    private void throwForInvalidIndexPair(int origin, int destination){
        if(matrix.length < origin || origin <= 0){
            throw new IndexOutOfBoundsException("Origin index " + origin + " is not in [1," + matrix.length + "] when" +
                " accessing od matrix.");
        }
        if(matrix.length < destination || destination <= 0){
            throw new IndexOutOfBoundsException("Destination index " + origin + " is not in [1," + matrix.length + "]" +
                "  when accessing od matrix.");
        }
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        builder.append("Full OD:\n");
        for(ODPair odPair : getODPairs()){
            builder.append(odPair).append("\n");
        }
        return builder.toString();
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        FullOD fullOD = (FullOD) o;

        return Arrays.deepEquals(matrix, fullOD.matrix);
    }

    @Override
    public int hashCode() {
        return Arrays.deepHashCode(matrix);
    }
}

File: src/core/python/core/model/impl/fullOD.py
from typing import List

import numpy

from core.model.od import OD, ODPair
from core.exceptions.index_exceptions import IndexOutOfBoundsException


class FullOD(OD):
    """
    Implements the od interface using a two dimensional numpy array. Therefore
    accessing the od pairs by index is cheap but getting nonempty od-pairs by
    getODPairs may be expensive.
    :param size     the size of the matrix
    """

    def __init__(self, size: int):
        self.matrix = numpy.zeros(shape=(size, size), dtype=float)

    def raise_for_invalid_index_pair(self, origin: int, destination: int):
        if self.matrix.shape[0] < origin or origin <= 0:
            exception_key = "Origin"
            raise IndexOutOfBoundsException(exception_key, origin,
                                            len(self.matrix))
        if self.matrix.shape[1] < destination or destination <= 0:
            exception_key = "Destination"
            raise IndexOutOfBoundsException(exception_key, origin,
                                            len(self.matrix))

    def getValue(self, origin: int, destination: int) -> float:
        self.raise_for_invalid_index_pair(origin, destination)
        return self.matrix[origin - 1][destination - 1]

    def setValue(self, origin: int, destination: int, new_value: float) -> None:
        self.raise_for_invalid_index_pair(origin, destination)
        self.matrix[origin - 1][destination - 1] = new_value

    def computeNumberOfPassengers(self) -> float:
        sum_value = 0
        for value in numpy.nditer(self.matrix):
                sum_value += value
        return sum_value

    def getODPairs(self) -> List[ODPair]:
        all_od_pairs = []
        it = numpy.nditer(self.matrix, flags=["multi_index"])
        while not it.finished:
            if not numpy.isclose(it[0], 0):
                all_od_pairs.append(ODPair(it.multi_index[0]+1, it.multi_index[1]+1, it[0]))
            it.iternext()
        return all_od_pairs

    def __str__(self) -> str:
        return "Full OD:\n" + "\n".join([str(pair) for pair in self.getODPairs()])

    def __eq__(self, other) -> bool:
        if not isinstance(other, FullOD):
            return False
        return numpy.allclose(self.matrix, other.matrix)

    def __ne__(self, other):
        return not self.__eq__(other)

    def __hash__(self):
        return hash(self.matrix)
